File: 0031-next-permutation/0031-next-permutation.py
class Solution:
    def nextPermutation(self, nums: List[int]) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """

        n = len(nums)
        i = n - 2 
        while i >= 0:
            if nums[i] < nums[i + 1]:
                for j in range(n-1, i, -1):
                    
                    if nums[j] > nums[i]:
                        nums[i], nums[j] = nums[j], nums[i]
                        # print(nums)
                        break
                
                break
            i -= 1
        
        self.reverse(nums, i + 1, n - 1)            
    
    def reverse(self,nums, start, end):
        
        while start < end:
            nums[start], nums[end] = nums[end], nums[start]
            start += 1
            end -= 1
    

File: 0031-next-permutation/0031-next-permutation.java
class Solution {
    public void nextPermutation(int[] nums) {
        
        int n = nums.length;
        int i = n - 2;
        while(i >= 0){
            if(nums[i] < nums[i + 1]){
                for(int j = n - 1; j > i; j--){
                    if(nums[j] > nums[i]){
                        swap(nums, i, j);
                        break;
                    }
                }
                break;
            }
            i--;
        }
        reverse(nums, i + 1, n- 1);
    }

    private void swap(int[] nums, int left, int right){

        int temp = nums[left];
        nums[left] = nums[right];
        nums[right] = temp;
    }

    private void reverse(int[] nums, int start, int end){

        while(start < end){
            swap(nums, start, end);
            start++;
            end--;
        }

    }
}